Url handling, data flow, and shuffle

avertin opened the pull request:
- When the user clicks "Go", update the url to include the category/location
- If the user navigates to a url that already contains a category/location, populate the dropdown menu with it
- Handle edge case where the url params don't match any of the categories or locations

I'll hold off on merging this PR until https://github.com/avertin/fund-justice/pull/5 is merged to avoid merge conflicts.  Once that PR is merged, I can also address  `// TODO figure out how to initialize these while the data is loading`

![urls](https://user-images.githubusercontent.com/11411307/84331239-87a0b600-ab3e-11ea-97da-7514e201ae86.gif)

agatheblues commented on src/App.jsx:
```diff
@@ -10,7 +10,7 @@ export default class App extends React.Component {
         <Header/>
         <Switch>
-          <Route exact path="/" component={Home} />
+          <Route exact path={['/', '/:category/:location']} component={Home} />
```
I see you've already did some work there, so feel free to argue against changing the approach if you think this could be changed later / is not so relevant 🙉 I think a more idiomatic approach would be to use query params rather than the url path, for those reasons:

- `category` is not really a 'parent' of `location` (in rest standards, the url path tells something about how resources are nested)
- these are filters/search terms
- if in the future we would allow to search by multiple states or categories, the url path would look funny 🍕 
- and most of all, we might run into problems of ordering for the router: if I choose a location first or if I choose the category first we will have to cleverly insert those parts in the url in the correct order to make the router happy, while query params are forgiving of the order.

So, two options:

*Opt1: we keep the params in url path*

-> In that case I would rather have a url shaped like this `/category/:category_id/location/:location_id`

*Opt2: we switch to query params*

So something like `{host}?location=NY&category=BAIL`
-> In that case, you can use this super cool [package](https://www.npmjs.com/package/query-string) that I have added to the repo in #3 to handle the search query param in the `sheet.best` calls

avertin replied:
Great suggestion!! Thanks for pointing this out. I went with the query params option.
See the commit called "PR Feedback: use query in url"

avertin commented:
@agatheblues this PR got a little bit out of hand. It includes a few more changes than just the url handling. I tried breaking up the commits by category. I added notes for things that I was hoping you could review and for outliers in the change relative to the commit category.

Url handling --

- Commit 1: [Include category/location selections in the url](https://github.com/avertin/fund-justice/pull/6/commits/b3f7eb14d4f249b39bfb5d0323e12436b6db620a)
- Commit 2: [Get url logic working after rebase](https://github.com/avertin/fund-justice/pull/6/commits/cf05ff821ae3215b92ed82c912ec5ee25eba8805). Note: After rebasing I understood why you were accessing object properties with `['propertyName']`. That was the way the api returned the data. To make life a little easier, I just reshaped the data to play nice with the object types we wanted in the components which was more clear after adding `PropTypes.shape`. In my opinion, the props should be pretty ignorant to the shape of the data from the api. It's also nice to keep the logic for reshaping the data right next to the place where the api returns it. So once the api returns the `rawData`, I just did a quick `rawData.map` to get an array that matches what the component wants.  Let me know if this approach makes sense. 
- Commit 8: [PR Feedback: use query in url](https://github.com/avertin/fund-justice/pull/6/commits/1575e7964ae6ada35983cd95f3b8e4c484261147)

Data flow / handling --
- Commit 3: [Add shape to object PropTypes](https://github.com/avertin/fund-justice/pull/6/commits/5152bcb11ec0e4ec150847237dd37571851a0235)
- Commit 4: [Wire up rest of data flow](https://github.com/avertin/fund-justice/pull/6/commits/0b6594eb5da07d4bb1763fbd115d6eea070c7197)
- Commit 7: [Wire up data flow part 2](https://github.com/avertin/fund-justice/pull/6/commits/93eda019bdb22b4f519463048e37decc0b2073a1)
- Commit 9: [Move currentOption state out of DropdownMenu](https://github.com/avertin/fund-justice/pull/6/commits/9acfa2d305e2adccd8b88a9761f017746ac68343) Note: it would be great if you could double check this looks right. 

Api --
- Commit 5: [Check response for failures](https://github.com/avertin/fund-justice/pull/6/commits/4e39ed9a2a77998da51296c74513ce7f986f561d). Note: When I was getting the 403, I noticed that the catch block for fetch wasn't being hit. This is expected but was leading to a crash since the component assumed it got a valid response back. As a result, we need to check the response status and throw an error if we don't get a 200. 
- Commit 6: [Implement shuffle](https://github.com/avertin/fund-justice/pull/6/commits/36c1acd4d6c6c2adcad62534586e526f89684224). Note: since I was having trouble with the API, I added some mock data to the repository so we can use it for UI debugging when we don't actually need to hit the server. I hope this will also slow down the rate that we hit the request limit for the api. 

Shuffle -- 
- Commit 6: [Implement shuffle](https://github.com/avertin/fund-justice/pull/6/commits/36c1acd4d6c6c2adcad62534586e526f89684224). Mock data stuff aside, this also contains the actual implementation for shuffle.

agatheblues commented on src/pages/Home.jsx:
```diff
@@ -1,34 +1,46 @@
+        <CardGrid
+          title={this.state.currentOptions.categoryOption.label}
+          categoryType={this.state.currentOptions.categoryOption.value}
+          location={this.state.currentOptions.locationOption.label}
```
Does the user have to select both a category *and* a location to click 'Go'? If not: one of `[categoryOption, locationOption]` might be null. Unless we always set a default like 'all categories' and 'all locations' (which from the api perspective translates to: no search parameters, but in the front-end would translate to real options)

avertin replied:
With the current implementation, yes. We default both of them to the first value returned by the api. Let's circle back with @rmakk for some design input on how to handle the defaults

agatheblues commented on src/components/OptionPicker.jsx:
```diff
@@ -1,48 +1,176 @@
+  handleShuffle = () => {
+    if (this.locationDropdownRef.current && this.categoryDropdownRef.current) {
+      const randomLocationOption = this.locationDropdownRef.current.shuffle();
+      const randomCategoryOption = this.categoryDropdownRef.current.shuffle();
```
Not sure I get why the `shuffle()` needs to happen in the children ref, since it's pushing a new state in the parent (here) that should update the dropdowns? We have access to the data at this level, so we could just run a shuffle on `this.state.categories.data` and `this.state.locations.data` here, assign the option to the state, and keep the dropdown 'dumb'. What do you think? Am I overlooking something?

avertin replied:
Ah yes that's a very good point. I think I put it in the child component before lifting the state out of DropdownMenu. I'll move it to OptionPicker

agatheblues commented on src/components/DropdownMenu.jsx:
```diff
@@ -49,18 +46,33 @@ export default class DropdownMenu extends React.Component {
+      // Keep shuffling until we get an index that is different than the current index.
+      while (randomIndex === currentIndex) {
+        randomIndex = Math.floor(Math.random() * this.props.options.length);
+      }
```
To prevent having a `while` loop, we could do:

```
rest = [...options.slice(0, currentIndex), ...options.slice(currentIndex + 1, options.length)] // remove current option from list
return rest[Math.floor(Math.random() * rest.length]
```

avertin replied:
Good idea 🙂

agatheblues commented on src/components/DropdownMenu.jsx:
```diff
@@ -8,29 +8,26 @@ export default class DropdownMenu extends React.Component {
-            {options.map(option => {
-              if (option[value] !== this.state.currentOption[value]) {
+            {this.props.options.map(option => {
+              if (option.value !== this.props.currentOption.value) {
```
Neat ✨

agatheblues commented on src/components/CardGrid.jsx:
```diff
@@ -22,9 +42,32 @@ export default class CardGrid extends React.Component {
+        category_type: this.props.categoryType,
+        state: this.props.location,
+      };
+    }
```
I _think_™️ that here you could eventually do

```
const requestFilters = { category_type: this.props.categoryType,  state: this.props.location};
```

right away, since in the `getFunds()` method we do `const params = qs.stringify(filters, { skipNull: true });` (so it would skip null anyways, and then we don't have to care whether the filters are null or not) 🤔

avertin replied:
Good call!

avertin commented:
@agatheblues thanks for all the feedback, it should all be addressed now, so I'm going to go ahead and merge this one.
